Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+
+    fileprivate enum ResponseState {
+        enum StreamControlState {
+            case downstreamHasDemand
+            case readEventPending
+            case waiting
+        }
+
+        case initialized
+        case receivingBody(StreamControlState)
+        case endReceived
+    }
+
+    enum Action {
+        enum NextMessageToSend {
+            case sendEnd
+            case startBodyStream
+        }
+
+        case verifyRequest
+
+        case sendRequestHead(HTTPRequestHead, startBody: Bool, startReadTimeoutTimer: TimeAmount?)
+        case sendBodyPart(IOData)
+        case sendRequestEnd(startReadTimeoutTimer: TimeAmount?)
+
+        case pauseRequestBodyStream
+        case resumeRequestBodyStream
+
+        case forwardResponseHead(HTTPResponseHead)
+        case forwardResponseBodyPart(ByteBuffer, resetReadTimeoutTimer: TimeAmount?)
+        case forwardResponseEnd(readPending: Bool, clearReadTimeoutTimer: Bool)
+
+        case failRequest(Error, closeStream: Bool)
+
+        case read
+        case wait
+    }
+
+    private var state: State = .initialized
+
+    private var isChannelWritable: Bool
+    private let idleReadTimeout: TimeAmount?
+
+    init(isChannelWritable: Bool, idleReadTimeout: TimeAmount?) {
+        self.isChannelWritable = isChannelWritable
+        self.idleReadTimeout = idleReadTimeout
+    }
+
+    mutating func writabilityChanged(writable: Bool) -> Action {
+        self.isChannelWritable = writable
+
+        switch self.state {
+        case .initialized,
+             .finished,
+             .failed:
+            return .wait
+
+        case .running(.verifyRequest, _), .running(.endSent, _):
+            return .wait
+
+        case .running(.streaming(let expectedBody, let sentBodyBytes, producer: .paused), let responseState):
+            if writable {
+                let requestState: RequestState = .streaming(
+                    expectedBodyLength: expectedBody,
+                    sentBodyBytes: sentBodyBytes,
+                    producer: .producing
+                )
+
+                self.state = .running(requestState, responseState)
+                return .resumeRequestBodyStream
+            } else {
+                // no state change needed
+                return .wait
+            }
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, producer: .producing), let responseState):
+            if !writable {
+                let requestState: RequestState = .streaming(
+                    expectedBodyLength: expectedBodyLength,
+                    sentBodyBytes: sentBodyBytes,
+                    producer: .paused
+                )
+                self.state = .running(requestState, responseState)
+                return .pauseRequestBodyStream
+            } else {
+                // no state change needed
+                return .wait
+            }
+        }
+    }
+
+    mutating func readEventCaught() -> Action {
+        return .read
```

Comment:
Why do we unconditionally read? Does this not just apply to the `running` state?